lib/pbench/test/unit/server/test_shell_cli.py

```diff
@@ -70,124 +153,75 @@ def test_find_the_unicorn(monkeypatch, make_logger):
+        state.
+        """
+
+        pid_sequence = [0, 0, 2, 0, 1, 10, 0]
+        pid_index = [0]  # The list allows skipping 'global' in the function
+
+        def fake_waitpid(pid: int, flags: int) -> tuple[int, int]:
+            assert pid == 0
+            assert flags == os.WNOHANG
+            time.sleep(0.2)
+            retval = pid_sequence[pid_index[0]]
+            if pid_index[0] < len(pid_sequence) - 1:
+                pid_index[0] += 1
+            return retval, 1
+
+        """Test startup and shutdown of the indexers"""
+        monkeypatch.setattr("pbench.cli.server.shell.subprocess.Popen", FakePopen)
+        monkeypatch.setattr("pbench.cli.server.shell.Event", FakeEvent)
+        monkeypatch.setattr("pbench.cli.server.shell.os.waitpid", fake_waitpid)
+
+        indexer = shell.PbenchIndexer(
+            bin_dir=Path("/bin"), cwd=Path("/cwd"), logger=make_logger
         )
-
-        assert ret_val == 1
-        assert os.environ["PATH"] == "bindir:a:b", f"PATH='{os.environ['PATH']}'"
-        assert commands == [["pbench-create-crontab", "/tmp"]]
-        assert unlink_record == ["unlink"]
-
-    @staticmethod
-    def test_generate_crontab_if_necessary_crontab_failed(monkeypatch, make_logger):
-        monkeypatch.setenv("PATH", "a:b")
-
-        unlink_record = []
-
-        def unlink(*args, **kwargs):
-            unlink_record.append("unlink")
-
-        monkeypatch.setattr(Path, "unlink", unlink)
-
-        commands = []
-
-        def run(args, cwd: Optional[str] = None) -> subprocess.CompletedProcess:
-            commands.append(args)
-            ret_val = 1 if args[0] == "crontab" else 0
-            return subprocess.CompletedProcess(args, ret_val)
-
-        # We don't have an existing crontab
-        monkeypatch.setattr(Path, "exists", exists_false)
-        monkeypatch.setattr(subprocess, "run", run)
-
-        ret_val = shell.generate_crontab_if_necessary(
-            "/tmp", Path("bindir"), "cwd", make_logger
-        )
-
-        assert ret_val == 1
-        assert os.environ["PATH"] == "bindir:a:b", f"PATH='{os.environ['PATH']}'"
-        assert commands == [
-            ["pbench-create-crontab", "/tmp"],
-            ["crontab", "/tmp/crontab"],
-        ]
-        assert unlink_record == ["unlink"]
+        assert indexer.indexer is None
+        assert indexer.reindexer is None
+
+        indexer.start()
+        assert not indexer.shutoff.flag
+        assert indexer.reaper.isDaemon()
+        assert len(FakePopen.processes) == 2
+        assert FakePopen.processes[1].cmd == "python3 /bin/pbench-index"
+        assert FakePopen.processes[2].cmd == "python3 /bin/pbench-index --re-index"
+        assert indexer.indexer.running
+        assert not indexer.indexer.waited
+        assert indexer.reindexer.running
+        assert not indexer.reindexer.waited
+
+        while pid_index[0] < len(pid_sequence) - 1:
+            time.sleep(0.1)
```

Comment:
Any particular reason for waiting `0.1`s?  And, is there supposed to be a relationship between the wait here and the one in `fake_waitpid()`?  (E.g., should one be calculated from the other?)

More importantly, this loop looks like it will hang if `pid_index[0]` stops progressing.  It would be good to have a maximum iteration count on this loop or an assertion that the current value of `pid_index[0]` is greater than it was on the last iteration (assuming that we can make the respective `sleep()`s produce that effect), so that we don't hang the whole test system if things fail here.